cmd/go-filecoin/init.go

```diff
@@ -180,44 +219,63 @@ func loadGenesis(ctx context.Context, rep repo.Repo, sourceName string) (consens
 		}
 		source = file
 	}
-	defer func() { _ = source.Close() }()
+	return source, nil
+}
 
+func extractGenesisBlock(source io.ReadCloser, rep repo.Repo) (*block.Block, error) {
 	bs := blockstore.NewBlockstore(rep.Datastore())
 	ch, err := car.LoadCar(bs, source)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(ch.Roots) != 1 {
-		return nil, fmt.Errorf("expected car with only a single root")
+	// need to check if we are being handed a car file with a single genesis block or an entire chain.
+	bsBlk, err := bs.Get(ch.Roots[0])
+	if err != nil {
+		return nil, err
+	}
+	cur, err := block.DecodeBlock(bsBlk.RawData())
+	if err != nil {
+		return nil, err
 	}
 
-	gif := func(cst *hamt.CborIpldStore, bs blockstore.Blockstore) (*block.Block, error) {
-		var blk block.Block
-
-		if err := cst.Get(ctx, ch.Roots[0], &blk); err != nil {
-			return nil, err
+	// the root block of the car file has parents, this file must contain a chain.
+	var gensisBlk *block.Block
+	if !cur.Parents.Equals(block.UndefTipSet.Key()) {
+		// walk back up the chain until we hit a block with no parents, the genesis block.
+		for !cur.Parents.Equals(block.UndefTipSet.Key()) {
+			bsBlk, err := bs.Get(cur.Parents.ToSlice()[0])
+			if err != nil {
+				return nil, err
+			}
+			cur, err = block.DecodeBlock(bsBlk.RawData())
+			if err != nil {
+				return nil, err
+			}
 		}
 
-		return &blk, nil
-	}
-
-	return gif, nil
-}
+		gensisBlk = cur
 
-func getNodeInitOpts(peerKeyFile string) ([]node.InitOpt, error) {
-	var initOpts []node.InitOpt
-	if peerKeyFile != "" {
-		data, err := ioutil.ReadFile(peerKeyFile)
-		if err != nil {
-			return nil, err
+		// the root cids of the car file are a tipset, they form the head of a chain
+		var heads []*block.Block
```

Comment:
Because this looks redundant at first glance (why not just return the cids?) please leave a comment explaining that the block data and tipset construction is needed to sort the cids in the correct order.

Reply:
You are right, this is redentant -- will remove and just log the CID's.